core/arch/arm/kernel/ree_fs_ta.c

```diff
@@ -112,6 +118,70 @@ static TEE_Result rpc_load(const TEE_UUID *uuid, struct shdr **ta,
 	return res;
 }
 
+static TEE_Result check_update_version(struct shdr_bootstrap_ta *hdr)
+{
+	struct shdr_bootstrap_ta hdr_entry = {0};
+	const struct tee_file_operations *ops = NULL;
+	struct tee_file_handle *fh = NULL;
+	TEE_Result res = TEE_SUCCESS;
+	bool entry_found = false;
+	size_t len = 0;
+	uint32_t i = 0;
+	struct tee_pobj pobj = {
+		.obj_id = (void *)ta_ver_db_obj_id,
+		.obj_id_len = sizeof(ta_ver_db_obj_id)
+	};
+
+#ifdef CFG_RPMB_FS
+	ops = tee_svc_storage_file_ops(TEE_STORAGE_PRIVATE_RPMB);
+#else
+	ops = tee_svc_storage_file_ops(TEE_STORAGE_PRIVATE);
+#endif
+
+	res = ops->open(&pobj, NULL, &fh);
+	if (res == TEE_ERROR_ITEM_NOT_FOUND)
+		res = ops->create(&pobj, false, NULL, 0, NULL, 0, NULL, 0, &fh);
+
+	if (res != TEE_SUCCESS)
+		return res;
+
+	for (i = 0; i < ta_ver_db_idx; i++) {
+		len = sizeof(hdr_entry);
+
+		res = ops->read(fh, i * len, &hdr_entry, &len);
+		if (res != TEE_SUCCESS && len != sizeof(hdr_entry))
+			goto out_ver;
```

Comment:
> If len isn't sizeof(hdr_entry) res need to be updated.

Ok.

> Perhaps some kind if recovery should be considered since it's likely to be persistent.

I am not sure about recovery that you are looking here. As we can't delete a particular entry from database which may itself lead to TA rollback.